elastic/kibana
# [Monitoring] Cluster state watch to Kibana alerting

Builds upon https://github.com/elastic/kibana/pull/54306
Relates to https://github.com/elastic/kibana/issues/42960

This PR migrates the next cluster alert, the cluster state one.

Most of the changes in the PR are due to the initial work not working that well for multiple alerts, so some refactoring was necessary.

In particular, the way in which we tell the client what to show in the alert in the UI changed. The original implementation didn't support what we needed to do with other alerts so that's been changed and hopefully makes sense. It's a bit of a "generic" approach and I am welcome to any better approaches here.

## Testing

- Set `KIBANA_ALERTING_ENABLED` to `true` in both `legacy/plugins/monitoring/common/constants.js` and `plugins/monitoring/common/constants.ts`
- Navigate to the monitoring app and follow the instructions in the alerting callout at the top to enable Kibana alerting
- If you do not see an alert in the UI yet, create an index using dev tools and go back to the monitoring UI and you should see one about replica shards.

## Screenshots
### Creation
<img width="2039" alt="Screen Shot 2020-03-30 at 2 30 32 PM" src="https://user-images.githubusercontent.com/56682/77948203-0c479c00-7293-11ea-9d60-2364ec7ad302.png">
<img width="1079" alt="Screen Shot 2020-03-30 at 2 30 55 PM" src="https://user-images.githubusercontent.com/56682/77948225-15386d80-7293-11ea-862d-882dceba9785.png">


### Post creation
<img width="2045" alt="Screen Shot 2020-03-30 at 1 45 09 PM" src="https://user-images.githubusercontent.com/56682/77944292-b2dc6e80-728c-11ea-9872-3c397db30921.png">

## TODO
- [ ] This cluster alert has a link to the indices page with an action label that indicates you can do something about it. Should we change this since you can't really do anything on that page? cc @sorantis
